text/0000-location.md

```diff
@@ -0,0 +1,421 @@
+Sawtooth's Sabre smart contract engine running via Hyperledger Transact, which
+covers all currently implemented DLT backends in Grid.
+
+Locations are managed by submitting transactions to Hyperledger Grid, which will
+process them with the Grid Location smart contract. The following transactions 
+are supported:
+- LocationCreate - create a Location and store it in state.
+- LocationUpdate - update (replace) the properties of a Location already in state.
+- LocationDelete - remove a Location from state.
+ 
+_*Disclaimer: LocationDelete is a potentially hazardous operation and needs to 
+be done with care._
+
+
+## Permissions
+[permissions]: #permissions
+
+Creation of GS1 locations is restricted to agents acting on behalf of the 
+organization that is associated with the GS1 company prefix encoded in the GLN. 
+Like Grid Product, an organization must have the GS1 company prefix in the 
+“gs1_company_prefixes” metadata field for its Pike organization. When a 
+location is created, its owning organization is stored with the location in an 
+“owner” field.
+
+Updates to locations are restricted to agents acting on behalf of the 
+organization stored in the location’s owner field. Only property fields can be 
+updated. They are updated by providing a complete and updated list of the 
+properties, and overwrites the entire list of property values. Location_id, 
+location_namespace, and owner fields are immutable.
+
+Deletion of locations is restricted to agents acting on behalf of the 
+organization in the location’s owner field. A setting will turn off deletion 
+entirely, with the intent that it could be enabled/disabled by a Grid 
+administrator. Disabling delete is useful because external systems may have 
+references to these locations and deleting them could leave dangling references.
+
+
+## GLN Validation
+[gln validation]: #glnvalidation
+
+GLN validation can reduce the opportunity for invalid data to enter the system.
+At a minimum, when GLNs are entered in clients and later processed in smart
+contracts, they will need to conform to the GLN format and the check digit
+should be correct. See [check digit calculator](https://www.gs1us.org/tools/check-digit-calculator "check digit calculator").
+
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## State
+[state]: #state
+
+### Location Representation
+
+The primary object stored in state is “Location”, which consists of a 
+location_id (GS1 GLN), an owner (org_id compatible with Pike), and a list of 
+property name/value pairs. The properties available are defined by the Grid 
+Location Schema Transaction Family and are restricted to the fields and rules of
+the GS1 Location Property Schema. Transactions which are responsible for setting
+location state values must ensure that the properties conform with the 
+requirements of the GS1 Location Property Schema.
+
+    message Location { 
+        enum LocationNamespace { 
```

Comment:
Right now this is proposing an enum that is basically the same as the enum of ProductNamespace. 

Say we had a system that used Products and Locations within a GS1 context, and at a later time we wanted to implement a different namespace for another open standard. This would require us to implement add a new enum value to each smart contract that has similar functionality. Further, if the standard was only relevant at the time for a subset of the namespaces, inconsistency could develop between these enums.

What would folks think about implementing a common enum (say `RecordType`) to be used by any smart contract that has similar functionality? This would also get around the "LocationType" conflation discussed above. With this approach, it would be up to the smart contracts to handle validation of other fields (such as location_id or product_id) based on this enum.

Reply:
I think that is a good idea

Reply:
Even though we have 'GS1' in the both currently does not mean the structs would necessarily continue to be equivalent. (Not all standards will cover every smart contract, etc. -- a standard covering locations might not cover products, or the reverse, etc.)

Even if they were always going to be the same, this is still the cleanest way to define them (as distinct).